### ISSUE TITLE
Odoo 9C: Overriding the method & its "depends" decorator [Bug] 

### ISSUE THREAD

**Issue by deepa4lok** [odoo/odoo#12459]
Impacted versions: Odoo 9 C

Am trying to override a standard method, which involves adding an another dependency field into the api decorator.

However, I noticed, newly added field in the depends decorator is not taken into consideration while saving the record.

Ex: Say in Purchase Order, the method called "**_amount_all**" has **@api.depends** on **"order_line.price_total"**. Now in my custom module I have overridden that method, and included one of my dependency field to it, like

> @api.depends('order_line.price_total','myfield')

Though it seems to work, when I change the value in myfield, however while saving the record, method is not triggered to recompute...

So In other words, newly added dependency fields are NOT taken into consideration while saving the record, as a result, method computation is not at all triggered. Hence old value is retained.

I believe this must be a bug.
Anybody has faced this issue?

**Comment by mart-e** [comment]
You should not override a method to change it's triggers, it is not the expected behaviour.
If you want to add other checks, create a new `api.depends('myfield')` method (that may call another constraint method if needed).

**Comment by deepa4lok** [comment]
@mart-e I guess you didn't understand my concept very well thoroughly and certainly not fair marking the issue as invalid, I never mentioned I wanted to override a method just to add an additional trigger, my requirement is such a way that I needed one of the parent component to calculate the monetary details.

Let me repeat the same with an Example: along with the sum of **price_subtotal (purchase.order.line)** am adding **xyz_charges(purchase.order)**. Now when I change the value of xyz_charges, PurchaseOrder should reflect the same in its total, which is a normal expected behaviour. However it is not happening in this case. So can you please check again thoroughly, and get back to me?

P.S. I have achieved the same logic successfully in my custom module, but when am not able to do the same in Standard method by overriding it, then definitely it is a bug? Or if you got any better solution please educate me.

Please don't tell me, to achieve my requirement I should start creating 3 new custom fields i.e custom_amount_total, custom_amount_tax, custom_amount_untaxed to fulfil my requirement which is utterly ridiculous.

**Comment by deepa4lok** [comment]
Code:

> 
> class PurchaseOrder(models.Model):
>     _inherit = ['purchase.order']
> 
>     # Overridden:
>     @api.depends('order_line.price_total', 'xyz_charges', 'currency_id')
>     def _amount_all(self):
>         """
>         Overridden from Purchase Order
>         """
>        for order in self:
> 
>             order_currency = order.currency_id
>             ctx = {'date' : order.date_order}
> 
>             amount_untaxed = amount_tax = 0.0
>             for line in order.order_line:
>                 amount_untaxed += line.currency_id.with_context(ctx).compute(line.price_subtotal, order_currency)
>                 amount_tax += line.currency_id.with_context(ctx).compute(line.price_tax, order_currency)
> 
>             order.update({
>                 'amount_untaxed': order.pricelist_id.currency_id.round(amount_untaxed + order.xyz_charges),
>                 'amount_tax': order.pricelist_id.currency_id.round(amount_tax),
>                 'amount_total': amount_untaxed + amount_tax + order.xyz_charges,
>             })

> xyz_charges = fields.Float(string='XYZ Charges', digits=dp.get_precision('Product Price'))


> class PurchaseOrderLine(models.Model):
>     _inherit = 'purchase.order.line'

> currency_id = fields.Many2one("res.currency", related=False, string="Currency", readonly=False,
>                                   required=True, store=True)

Note: In my requirement, Lines also will have a currency which may be different from Parent Object i.e PurchaseOrder. Hence my calculation is depended solely on currency, and yes I have overridden the line currency field by nullifying the related property.

I hope this gives the full insight of what am trying to achieve.
Any thought on this is appreciated. Thanks

**Comment by mart-e** [comment]
Ok thanks for the example. Indeed this scenario should be covered, sorry about that.

However I just made a quick test and it seemed to work fine
```python
class PO(models.Model):
    _inherit = 'purchase.order'

    additional_fee = fields.Float()

    @api.depends('additional_fee')
    def _amount_all(self):
        super(PO, self)._amount_all()
        for order in self:
            order.amount_total += order.additional_fee
```

The amount_total field was correctly recomputed when either `order_line.price_total` or `additional_fee` was modified.

**Comment by deepa4lok** [comment]
@mart-e Yes this is the perfect example,appreciate your effort, can please you add currency_id and check as well...

As I mentioned earlier, 
case 1: When I change xyz_charges amount_total gets recomputed well,
case 2: When I change xyz_charges & currency_id, works fine.
**case 3: When only currency_id is changed, the amount_total doesn't change,** 

i.e. amount_total should be calculated based on the conversion rate of selected Currency. (because my line currency & order currency will be different)

P.S. concept of concatenation on depends is wonderful, thanks for sharing it..

**Comment by mart-e** [comment]
Do you mean you change the currency of the line or of the order?
You are adding a field `currency_id` on the `purchase.order.line` and your trigger of `_amount_all` is on the field `currency_id` of the `purchase.order`.

**Comment by deepa4lok** [comment]
On change of currency present in the Order.

Try the below Scenario:
Company Currency is EUR,

POL: price_total is 1000 INR.   (_as mentioned earlier POL currency is no more related to PO_)
PO : Currency is USD 
Hence PO Total will be **14.70 USD.**

Save this record. (_this is needed in order to face the scenario_)

Again re-edit & change the Order Currency alone to GBP.   (_Note: do not change any line value_)
Now the PO Total will reflect immediately as **10.74 GBP** (Which is correct)

However when you Save the record, the PO total retains as **14.70 GBP** (Which is nothing but the previously stored value of the Total, so re-compute trigger upon save is not been called here)

Please try to create this scenario, by following exactly the above steps, I hope you will get the clear picture of it...
Thanks.

**Comment by mart-e** [comment]
Sorry but github is not the right place to ask development advises.
If you can show (a scenario clear with a small POC) there is something wrong on the trigger mechanism, I can reopen but here it's more the logic of the calls than the depends mechanis.
I advise to check on [forum](https://www.odoo.com/forum/help-1/), [mailing lists](https://www.odoo.com/groups) or contact the support/partner.

**Comment by deepa4lok** [comment]
@mart-e well I never asked for development advises, infact I have achieved this particular functionality in my custom module, I have given the step by step procedure to re-create the issue, am sure you could have seen the bug yourself if you had followed it... 

Anyways No issues, Thanks for your effort, yes please re-open the issue, possibly somebody might find a way.
